buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/CopyRestApiPlugin.java

```diff
@@ -0,0 +1,136 @@
+ * spec and YAML based rest tests.
+ * </p>
+ * <strong>Rest API specification:</strong> <br>
+ * When the {@link CopyRestApiPlugin} has been applied the {@link CopyRestApiTask} will automatically copy the core Rest API specification
+ * if there are any Rest YAML tests present in source, or copied from {@link CopyRestTestsTask} output. X-pack specs must be explicitly
+ * declared to be copied.
+ * <br>
+ * <i>For example:</i>
+ * <pre>
+ * restResources {
+ *   restApi {
+ *     includeXpack 'enrich'
+ *   }
+ * }
+ * </pre>
+ * Will copy the entire core Rest API specifications (assuming the project has tests) and any of the the X-pack specs starting with enrich*.
+ * It is recommended (but not required) to also explicitly declare which core specs your project depends on to help optimize the caching
+ * behavior.
+ * <i>For example:</i>
+ * <pre>
+ * restResources {
+ *   restApi {
+ *     includeCore 'index', 'cat'
+ *     includeXpack 'enrich'
+ *   }
+ * }
+ * </pre>
+ * <br>
+ * <strong>Rest YAML tests :</strong> <br>
+ * When the {@link CopyRestApiPlugin} has been applied the {@link CopyRestTestsTask} will copy the Rest YAML tests if explicitly
+ * configured with `includeCore` or `includeXpack` through the `restResources.restTests` extension.
+ * <i>For example:</i>
+ * <pre>
+ * restResources {
+ *  restApi {
+ *      includeXpack 'graph'
+ *   }
+ *   restTests {
+ *     includeXpack 'graph'
+ *   }
+ * }
+ * </pre>
+ * Will copy any of the the x-pack tests that start with graph, and will copy the X-pack graph specification, as well as the full core
+ * Rest API specification.
+ *
+ * @see CopyRestApiTask
+ * @see CopyRestTestsTask
+ */
+public class CopyRestApiPlugin implements Plugin<Project> {
+
+    private static final String EXTENSION_NAME = "restResources";
+
+    @Override
+    public void apply(Project project) {
+        RestResourcesExtension extension = project.getExtensions().create(EXTENSION_NAME, RestResourcesExtension.class);
+
+        Provider<CopyRestTestsTask> copyRestYamlTestTask = project.getTasks()
+            .register("copyYamlTestsTask", CopyRestTestsTask.class, task -> {
+                task.includeCore.set(extension.restTests.getIncludeCore());
+                task.includeXpack.set(extension.restTests.getIncludeXpack());
+                task.coreConfig = project.getConfigurations().create("restTest");
+                if (BuildParams.isInternal()) {
+                    Dependency dependency = project.getDependencies()
+                        .project(Map.of("path", ":rest-api-spec", "configuration", "restTests"));
+                    project.getDependencies().add(task.coreConfig.getName(), dependency);
+
+                    task.xpackConfig = project.getConfigurations().create("restXpackTest");
+                    dependency = project.getDependencies().project(Map.of("path", ":x-pack:plugin", "configuration", "restXpackTests"));
```

Comment:
Let's avoid reassigning variables like this. It's a bit difficult to follow and opens up opportunity for nasty bugs vs using explicit variables.